### PR TITLE
🐛 handle nil pointer in clusterctl describe

### PR DESCRIPTION
**Pull request by bavarianbidi**
**What this PR does / why we need it**:

if a cluster as an empy bootstrap.configref in the machinedeployment (which is allowed), a nil pointer will raise if clusterctl describe cluster is called with --show-templates.

Also add a second type of a fake machinedeployment which has a secret reference in the boostrap object instead of a config ref

```
/*
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x19478c0]

goroutine 1 [running]:
sigs.k8s.io/cluster-api/cmd/clusterctl/client/tree.ObjectReferenceObject(0x0)
        sigs.k8s.io/cluster-api/cmd/clusterctl/client/tree/util.go:116 +0x40
sigs.k8s.io/cluster-api/cmd/clusterctl/client/tree.addMachineDeploymentToObjectTree({0x2153e90, 0xc000120008}, {0x215ebf8, 0xc0002ca770}, 0xc000c336c0, 0xc000c336c0?, 0x2171100?, 0xc0009e2200, {{0x7fffd7f57514, 0x3}, ...}, ...)
        sigs.k8s.io/cluster-api/cmd/clusterctl/client/tree/discovery.go:244 +0x35b
sigs.k8s.io/cluster-api/cmd/clusterctl/client/tree.Discovery({0x2153e90, 0xc000120008}, {0x215ebf8?, 0xc0002ca770?}, {0x7fffd7f574f3, 0xe}, {0x7fffd7f574eb, 0x4}, {{0x7fffd7f57514, 0x3}, ...})
        sigs.k8s.io/cluster-api/cmd/clusterctl/client/tree/discovery.go:145 +0x94c
sigs.k8s.io/cluster-api/cmd/clusterctl/client.(*clusterctlClient).DescribeCluster(0x0?, {{{0x0, 0x0}, {0x0, 0x0}}, {0x7fffd7f574f3, 0xe}, {0x7fffd7f574eb, 0x4}, {0x7fffd7f57514, ...}, ...})
        sigs.k8s.io/cluster-api/cmd/clusterctl/client/describe.go:91 +0x218
sigs.k8s.io/cluster-api/cmd/clusterctl/cmd.runDescribeCluster(0x0?, {0x7fffd7f574eb, 0x4})
        sigs.k8s.io/cluster-api/cmd/clusterctl/cmd/describe_cluster.go:154 +0x1d8
sigs.k8s.io/cluster-api/cmd/clusterctl/cmd.glob..func6(0x31aae60?, {0xc00018af80?, 0x8?, 0x8?})
        sigs.k8s.io/cluster-api/cmd/clusterctl/cmd/describe_cluster.go:105 +0x2d
github.com/spf13/cobra.(*Command).execute(0x31aae60, {0xc00018af00, 0x8, 0x8})
        github.com/spf13/cobra@v1.7.0/command.go:940 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0x31ad980)
        github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.7.0/command.go:992
sigs.k8s.io/cluster-api/cmd/clusterctl/cmd.Execute()
        sigs.k8s.io/cluster-api/cmd/clusterctl/cmd/root.go:105 +0x25
main.main()
        sigs.k8s.io/cluster-api/cmd/clusterctl/main.go:27 +0x17
*/
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by killianmuldoon** [comment]
Second time I've seen this flake where the env cleanup fails :thinking:

**Review by ykakarap — COMMENTED** [review]
The nil check makes sense. 

I will do another pass and will take a closer look at the updated tests.

**Comment by bavarianbidi** [comment]
> The nil check makes sense.
> 
> I will do another pass and will take a closer look at the updated tests.

Thank you.

In the end, i just copied the `FakeMachineDeployment` generation to create another md where the `bootstrap.configref` is nil and a secret is used.
Wherever the origin `FakeMachineDeployment` was used in the test, i've added an additional test-case (or extended an existing one) to make use of the newly generated fake md as well.

File: cmd/clusterctl/client/cluster/mover_test.go [hunk 1]
```diff
@@ -248,6 +248,57 @@ var moveTests = []struct {
...
+		fields: moveTestsFields{
+			objs: test.NewFakeCluster("ns1", "cluster1").
+				WithMachineDeploymentsWithoutBootstrapConfigRef(
+					test.NewFakeMachineDeployment("md1").
+						WithMachineSets(
+							test.NewFakeMachineSet("ms1").
+								WithMachines(
+									test.NewFakeMachine("m1"),
+									test.NewFakeMachine("m2"),
+								),
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover_test.go:262]: This seems a little bit of topic for this PR, but I'm not sure if I got right what we intend with "static bootstrap secret", because If I'm not wrong this is going to create a bootstrap config object (which I can see below at L289-290), which usually does not exist when a bootstrap secret is provided.
The same applies to the test case for the object graph

File: cmd/clusterctl/internal/test/fake_objects.go [hunk 2]
```diff
@@ -43,16 +43,17 @@ import (
...
-	machines              []*FakeMachine
-	withCloudConfigSecret bool
-	withCredentialSecret  bool
-	topologyClass         *string
+	namespace                                   string
+	name                                        string
+	controlPlane                                *FakeControlPlane
+	machinePools                                []*FakeMachinePool
+	machineDeployments                          []*FakeMachineDeployment
+	machineDeploymentsWithoutBootstrapConfigRef []*FakeMachineDeployment
```
Comment (fabriziopandini) [inline cmd/clusterctl/internal/test/fake_objects.go:51]: Having two sets of machine deployments seems counter-intuitive, can we solve the same use case by adding an attribute at MD deployment level, so we can Do test.NewFakeMachineDeployment("md02").WithStaticBootstrapConfig.

Then also fake MS and fake Machines can detect the setting in the parent object and generate resources accordingly

File: cmd/clusterctl/client/tree/discovery_test.go [hunk 3]
```diff
@@ -39,7 +39,7 @@ func clusterObjectsWithResourceSet() []client.Object {
 				),
 		).
 		WithMachineDeployments(
-			test.NewFakeMachineDeployment("md1").
+			test.NewFakeMachineDeployment("md01").
```
Comment (ykakarap) [inline cmd/clusterctl/client/tree/discovery_test.go:42]: Is this rename necessary?

File: cmd/clusterctl/client/tree/discovery_test.go [hunk 4]
```diff
@@ -417,22 +428,29 @@ func Test_Discovery(t *testing.T) {
...
-					"bootstrap.cluster.x-k8s.io/v1beta1, Kind=GenericBootstrapConfigTemplate, ns1/md1",
+				"virtual.cluster.x-k8s.io/v1beta1, Kind=TemplateGroup, ns1/md02": {
+					"infrastructure.cluster.x-k8s.io/v1beta1, Kind=GenericInfrastructureMachineTemplate, ns1/md02",
 				},
 				// MachineDeployment InfrastructureRef should be a leaf
-				"infrastructure.cluster.x-k8s.io/v1beta1, Kind=GenericInfrastructureMachineTemplate, ns1/md1": {},
+				"infrastructure.cluster.x-k8s.io/v1beta1, Kind=GenericInfrastructureMachineTemplate, ns1/md02": {},
 				// MachineDeployment BootstrapConfigRef should be a leaf
-				"bootstrap.cluster.x-k8s.io/v1beta1, Kind=GenericBootstrapConfigTemplate, ns1/md1": {},
+				"bootstrap.cluster.x-k8s.io/v1beta1, Kind=GenericBootstrapConfigTemplate, ns1/md01": {},
```
Comment (ykakarap) [inline cmd/clusterctl/client/tree/discovery_test.go:453]: Can you please move this below l.445 so that everything related `md01` is grouped together?

File: cmd/clusterctl/client/tree/discovery_test.go [hunk 5]
```diff
@@ -417,22 +428,29 @@ func Test_Discovery(t *testing.T) {
...
+					"virtual.cluster.x-k8s.io/v1beta1, Kind=TemplateGroup, ns1/md01",
+				},
+				// MachineDeployment TemplateGroup should have a BootstrapConfigRef and InfrastructureRef
+				"virtual.cluster.x-k8s.io/v1beta1, Kind=TemplateGroup, ns1/md01": {
+					"infrastructure.cluster.x-k8s.io/v1beta1, Kind=GenericInfrastructureMachineTemplate, ns1/md01",
+					"bootstrap.cluster.x-k8s.io/v1beta1, Kind=GenericBootstrapConfigTemplate, ns1/md01",
 				},
+				// MachineDeployment InfrastructureRef should be a leaf
+				"infrastructure.cluster.x-k8s.io/v1beta1, Kind=GenericInfrastructureMachineTemplate, ns1/md01": {},
 				// MachineDeployment TemplateGroup should have a BootstrapConfigRef and InfrastructureRef
```
Comment (ykakarap) [inline cmd/clusterctl/client/tree/discovery_test.go:446]:
```suggestion
				// MachineDeployment TemplateGroup using static bootstrap will only have InfrastructureRef
```

File: cmd/clusterctl/client/tree/discovery_test.go [hunk 6]
```diff
@@ -417,22 +428,29 @@ func Test_Discovery(t *testing.T) {
 				"cluster.x-k8s.io/v1beta1, Kind=Machine, ns1/cp1": {},
 				// Workers should have a machine deployment
 				"virtual.cluster.x-k8s.io/v1beta1, Kind=WorkerGroup, ns1/Workers": {
-					"cluster.x-k8s.io/v1beta1, Kind=MachineDeployment, ns1/md1",
+					"cluster.x-k8s.io/v1beta1, Kind=MachineDeployment, ns1/md01",
+					"cluster.x-k8s.io/v1beta1, Kind=MachineDeployment, ns1/md02",
 				},
 				// Machine deployment should have a group of machines (grouping) and templates group
-				"cluster.x-k8s.io/v1beta1, Kind=MachineDeployment, ns1/md1": {
+				"cluster.x-k8s.io/v1beta1, Kind=MachineDeployment, ns1/md01": {
```
Comment (ykakarap) [inline cmd/clusterctl/client/tree/discovery_test.go:435]: Can you please add a similar check for `md02`.

Please group all md01 checks and md02 checks together. The test become more readable that way. :)

File: cmd/clusterctl/internal/test/fake_objects.go [hunk 7]
```diff
@@ -531,10 +532,28 @@ func NewFakeInfrastructureTemplate(name string) *fakeinfrastructure.GenericInfra
 	}
 }
 
+func NewStaticBootstrapConfig() *clusterv1.Bootstrap {
```
Comment (ykakarap) [inline cmd/clusterctl/internal/test/fake_objects.go:535]:
```suggestion
func NewStaticBootstrapConfig(name string) *clusterv1.Bootstrap {
```

How about doing this and setting the `DataSecretName` to the passed argument. Would be nice to be able to control the value per FakeMachineDeployment instead of using the same for every MachineDeployment.

Also needs a change to `WithStaticBootstrapConfig` to take a name argument.

File: cmd/clusterctl/client/cluster/mover_test.go [hunk 8]
```diff
@@ -248,6 +248,59 @@ var moveTests = []struct {
...
+		wantMoveGroups: [][]string{
+			{ // group 1
+				"cluster.x-k8s.io/v1beta1, Kind=Cluster, ns1/cluster1",
+			},
+			{ // group 2 (objects with ownerReferences in group 1)
+				// owned by Clusters
+				"/v1, Kind=Secret, ns1/cluster1-ca",
+				"/v1, Kind=Secret, ns1/cluster1-kubeconfig",
+				"cluster.x-k8s.io/v1beta1, Kind=MachineDeployment, ns1/md1",
+				"bootstrap.cluster.x-k8s.io/v1beta1, Kind=GenericBootstrapConfigTemplate, ns1/md1",
```
Comment (ykakarap) [inline cmd/clusterctl/client/cluster/mover_test.go:276]: If we are using a statis bootstrap then there should not be any `GenericBootstrapConfigTemplate` in the output. Isn't it?
(Same applies for the 2 `GenericBootstrapConfig` entries below.)

Looks like this is because even if we are using a static bootstrap config we are creating and adding the BootstrapConfigTemplate and the associated BootstrapConfigs into the output object [here](https://github.com/ykakarap/cluster-api/blob/ba83907ee1749ea38652fca55cb99f5fa2014cc8/cmd/clusterctl/internal/test/fake_objects.go#L667). This should ideally be adjusted to reflect a case where we only have the associated secrets and not the BootstrapConfigs.

File: cmd/clusterctl/client/cluster/objectgraph_test.go [hunk 9]
```diff
@@ -810,6 +811,88 @@ var objectGraphsTests = []struct {
...
+					owners: []string{
+						"cluster.x-k8s.io/v1beta1, Kind=Machine, ns1/m1",
+					},
+				},
+				"bootstrap.cluster.x-k8s.io/v1beta1, Kind=GenericBootstrapConfig, ns1/m1": {
+					owners: []string{
+						"cluster.x-k8s.io/v1beta1, Kind=Machine, ns1/m1",
+					},
+				},
+				"bootstrap.cluster.x-k8s.io/v1beta1, Kind=GenericBootstrapConfigTemplate, ns1/md1": {
```
Comment (ykakarap) [inline cmd/clusterctl/client/cluster/objectgraph_test.go:883]: Same as the comments in `mover_test.go`.

File: cmd/clusterctl/internal/test/fake_objects.go (hunk 7 above)
Comment (bavarianbidi) [reply cmd/clusterctl/internal/test/fake_objects.go:535]: I had that in the beginning. As no test check on the name of the secret itself i decided to use some kind of a static secret.

The only string which is around is the name of the `machinedeployment` itself. We could either use this name as `DataSecretName` or  use it as prefix.

Comment (ykakarap) [reply cmd/clusterctl/internal/test/fake_objects.go:535]: IMHO it is better to have unique name rather than use a static name. Let's use the machinedeployment name as a prefix and either way I think you will still have to change `NewStaticBootstrapConfig ` so that it takes a `name string` argument and while calling it from inside `WithStaticBootstrapConfig` use pass `<md-name>-datasecret` as the argument value. 

Thoughts?

File: cmd/clusterctl/client/cluster/mover_test.go (hunk 8 above)
Comment (ykakarap) [reply cmd/clusterctl/client/cluster/mover_test.go]: If we are using a MachineDeployment/MachineSet with a static bootstrap (.bootstrap.configRef == nil) then no `KubedmConfig` would be created when the MachineSet creates the required Machines.

I am curious in the case you shared here was the KubeadmConfig created by the MachineDeployment/MachineSet while the MachineDeployment/MachineSet only has `dataSecretName` under `bootstrap` and no `configRef`? 🤔

Comment (bavarianbidi) [reply cmd/clusterctl/client/cluster/mover_test.go]: i'm very sorry for the confusion :see_no_evil: but i've turned wrong when i analyzed the `machineset|machinedeployment` where `.bootstrap.configRef == nil`.

the machine `anzu-2qpjh` is a `KCP` based machine where the config is different.

if i check the correct machine, the config is exactly as it's expected:

```yaml
spec:
  bootstrap:
    dataSecretName: anzu-md-ignition
  clusterName: anzu
  infrastructureRef:
    [...]
```

Comment (bavarianbidi) [reply cmd/clusterctl/client/cluster/mover_test.go]: > If we are using a MachineDeployment/MachineSet with a static bootstrap (.bootstrap.configRef == nil) then no `KubedmConfig` would be created when the MachineSet creates the required Machines.

@ykakarap i've changed the fake machine Obj implementation to make this case work. PTAL :pray:

**Comment by fabriziopandini** [comment]
thanks for this PR
/lgtm

@ykakarap @Jont828 for a final pass

**Comment by bavarianbidi** [comment]
@ykakarap / @Jont828 any chance to get this merged soon or do you want me to add something more?

**Comment by sbueringer** [comment]
Thank you very much. Very nice test coverage!!

/lgtm
/approve
